docs/structured-streaming-programming-guide.md

```diff
@@ -0,0 +1,888 @@
+    .format("console")
+    .option("checkpointLocation", checkpointDir)
+    .start();
+
+query.awaitTermination();
+{% endhighlight %}
+
+</div>
+<div data-lang="python"  markdown="1">
+
+{% highlight python %}
+query = wordCounts\
+    .writeStream\
+    .outputMode('complete')\
+    .format('console')\
+    .option('checkpointLocation', checkpointDir)\
+    .start()
+
+query.awaitTermination()
+{% endhighlight %}
+
+</div>
+</div>
+
+Now the streaming computation has started in the background, and the `query` object is a handle to that active streaming query. Note that we are also waiting for the query to terminate, to prevent the process from finishing while the query is active.
+To actually run this code, you can either compile your own Spark application, or simply run the example once you have downloaded Spark. We are showing the latter. You will first need to run Netcat (a small utility found in most Unix-like systems) as a data server by using
+
+    $ nc -lk 9999
+
+Then, in a different terminal, you can start the example by using
+
+<div class="codetabs">
+<div data-lang="scala"  markdown="1">
+
+    $ ./bin/run-example org.apache.spark.examples.sql.streaming.StructuredNetworkWordCount
+
+</div>
+<div data-lang="java"  markdown="1">
+
+    $ ./bin/run-example org.apache.spark.examples.sql.streaming.JavaStructuredNetworkWordCount 
+
+</div>
+<div data-lang="python"  markdown="1">
+    
+    $ ./bin/spark-submit examples/src/main/python/sql/streaming/structured_network_wordcount.py
+
+</div>
+</div>
+
+Then, any lines typed in the terminal running the netcat server will be counted and printed on screen every second. It will look something like the following.
+
+# Programming Model
+
+The key idea is in Structured Streaming is to treat a live data stream as a 
+table that is being continuously appended. This leads to a new stream 
+processing model that is very similar to a batch processing model. You will 
+express your streaming computation as standard batch-like query as on a static 
+table, and Spark runs it as an *incremental* query on the *unbounded* input 
+table. Let’s understand this model in more details.
+
+## Basic Concepts
+Consider the input data stream as the “Input Table”. Every data items that is 
+arriving on the stream is like a new row being appended to the Input Table.
+
+![Stream as a Table](img/structured-streaming-stream-as-a-table.png "Stream as a Table")
+
+A query on the input will generate the “Result Table”. Every trigger interval (say, every 1 second), new rows gets appended to the Input Table, which eventually updates the Result Table. Whenever the result table gets updated, we would want write the changed result rows to an external sink. 
+
+![Model](img/structured-streaming-model.png)
+
+The “Output” is defined as what gets written out to the external storage. The output can be defined in different modes 
+
+  - *Complete Mode* - The entire updated Result Table will be written to the external storage.
```

Comment:
Need to clarify whether this duplicates entire result table to external storage repeatedly, i.e. at every trigger interval